README.md
# concourse (the rubygem)

The `concourse` gem provides rake tasks to help you manage Concourse CI pipelines, jobs, and workers, to assist in running tasks with `fly execute`, and even run a local ephemeral deployment of Concourse on your development machine.

If you're not familiar with Concourse CI, you can read up on it at https://concourse-ci.org

Here's an example pipeline maintained by this gem:

![nokogiri-master](assets/nokogiri-master-pipeline.png)


## Contents

<!-- to update this TOC run `rake readme_toc` -->

<!-- toc -->

- [Basic Usage](#basic-usage)
  * [Install](#install)
  * [Add to your `Rakefile`](#add-to-your-rakefile)
  * [Set up your Concourse pipeline](#set-up-your-concourse-pipeline)
- [Real-world Examples](#real-world-examples)
- [Concourse pipeline configuration](#concourse-pipeline-configuration)
  * [ERB Templating](#erb-templating)
  * [YTT Templating](#ytt-templating)
  * [Mixing ERB and YTT Templating](#mixing-erb-and-ytt-templating)
  * [Secrets](#secrets)
  * [Multiple pipelines](#multiple-pipelines)
- [Configuration](#configuration)
  * [`directory`: Concourse subdirectory name](#directory-concourse-subdirectory-name)
  * [`fly_target`: Concourse `fly` target name](#fly_target-concourse-fly-target-name)
  * [`ytt`: boolean, or a YTT configuration directory name](#ytt-boolean-or-a-ytt-configuration-directory-name)
  * [`format`: Emit the final pipelines in `fly format-pipeline` canonical format](#format-emit-the-final-pipelines-in-fly-format-pipeline-canonical-format)
  * [`pipeline_erb_filename`: Pipeline filename](#pipeline_erb_filename-pipeline-filename)
  * [`secrets_filename`: Secrets filename](#secrets_filename-secrets-filename)
  * [`fly_args_`: fly command arguments](#fly_args_-fly-command-arguments)
- [Rake Tasks](#rake-tasks)
  * [Manage your Concourse pipeline](#manage-your-concourse-pipeline)
  * [List pipeline tasks](#list-pipeline-tasks)
  * [Run a Concourse task with `fly execute`](#run-a-concourse-task-with-fly-execute)
  * [Abort running builds](#abort-running-builds)
  * [Prune stalled concourse workers](#prune-stalled-concourse-workers)
- [Running a local ephemeral Concourse](#running-a-local-ephemeral-concourse)
  * [Spinning up a local deployment](#spinning-up-a-local-deployment)
  * [Doing things with the local deployment](#doing-things-with-the-local-deployment)
  * [Shutting down the local deployment](#shutting-down-the-local-deployment)
- [Contributing](#contributing)
- [License](#license)

<!-- tocstop -->

## Basic Usage

### Install

Add this line to your application's Gemfile:

```ruby
gem 'concourse'
```

and then run `bundle`.

You can also install it via `gem install concourse`.


### Add to your `Rakefile`

The primary functionality of the Concourse gem is to provide rake tasks. So, in your Rakefile:

``` ruby
require 'concourse'

Concourse.new("myproject").create_tasks!
```

### Set up your Concourse pipeline

``` sh
rake concourse:init
```

The `concourse:init` task will do a few different things for you:

1. create a subdirectory named `concourse` (or whatever you've configured with the `:directory` parameter)
2. create an empty Concourse pipeline file in that subdirectory named `<myproject>.yml` (or whatever you've configured with the `:pipeline_erb_filename` parameter)
3. ensure git will ignore your secrets file named `private.yml` (or whatever you've configured with the `:secrets_filename` parameter)


## Real-world Examples

It might be helpful to look at how other projects are using this gem. Here are a sample, ordered (approximately) from simplest to most complex:

* [`jbarnette/hoe-debugging`](https://github.com/jbarnette/hoe-debugging) has [one pipeline](https://github.com/jbarnette/hoe-debugging/blob/master/concourse/hoe-debugging.yml) that test multiple versions of MRI. ([see it live](https://ci.nokogiri.org/teams/flavorjones/pipelines/hoe-debugging))
* [`flavorjones/hoe-gemspec`](https://github.com/flavorjones/hoe-gemspec) has [one pipeline](https://github.com/flavorjones/hoe-gemspec/blob/master/concourse/hoe-gemspec.yml) that test multiple versions of MRI __and JRuby__. ([see it live](https://ci.nokogiri.org/teams/flavorjones/pipelines/hoe-gemspec))
* [`flavorjones/calendar-assistant`](https://github.com/flavorjones/calendar-assistant) has [one pipeline](https://github.com/flavorjones/calendar-assistant/blob/master/concourse/calendar-assistant.yml) that runs tests against `master` __and pull-requests__, uses __secrets__, and __integrates with Slack__. ([see it live](https://ci.nokogiri.org/teams/calendar-assistants/pipelines/calendar-assistant))
* [`flavorjones/mini_portile`](https://github.com/flavorjones/mini_portile) has [one pipeline](https://github.com/flavorjones/mini_portile/blob/master/concourse/mini_portile.yml) that run tests on both Linux __and Windows__. ([see it live](https://ci.nokogiri.org/teams/nokogiri-core/pipelines/mini_portile))
* [`flavorjones/chromedriver-helper`](https://github.com/flavorjones/chromedriver-helper) has __two pipelines__ that test [master](https://github.com/flavorjones/chromedriver-helper/blob/master/concourse/chromedriver-helper.yml) and [pull requests](https://github.com/flavorjones/chromedriver-helper/blob/master/concourse/chromedriver-helper-pr.yml). (see [Rakefile](https://github.com/flavorjones/chromedriver-helper/blob/master/Rakefile#L9-L12) config) (see [master](https://ci.nokogiri.org/teams/flavorjones/pipelines/chromedriver-helper) and [PR](https://ci.nokogiri.org/teams/flavorjones/pipelines/chromedriver-helper-pr) pipelines live)
* [`sparklemotion/nokogiri`](https://github.com/sparklemotion/nokogiri) has [multiple pipelines](https://github.com/sparklemotion/nokogiri/blob/master/concourse/nokogiri.yml) that __use `require` and `erbify_file` to share common elements__, and __integrate with IRC__. (see [master](https://ci.nokogiri.org/teams/nokogiri-core/pipelines/nokogiri) and [PR](https://ci.nokogiri.org/teams/nokogiri-core/pipelines/nokogiri-pr) pipelines live)


## Concourse pipeline configuration

### ERB Templating

Your Concourse pipeline configuration file, `<myproject>.yml` (or whatever you've configured with the `:pipeline_erb_filename` parameter), will be treated like an ERB template.

The concourse directory is added the `$LOAD_PATH` and so local ruby files may be `require`d if you so desire. Also note that you can include other yaml erb snippets with the method `erbify_file`.

An example using both of these features:

``` yaml
% require "common_prelude" #  will find "common_prelude.rb" in the concourse directory

resources:
<%= erbify_file "common_resources.yml" -%> #  will find this file in the concourse directory and recursively erbify it
  - name: resource_unique_to_this_pipeline
    type: git
    source:
      uri: <%= $common_prelude_defined_uri_prefix %>/username/projectname.git
      branch: master
```

(If you're unfamiliar with ERB and how you can mix Ruby into the document, you can [read about it here](https://www.stuartellis.name/articles/erb/).)


#### `RUBIES`

The ruby variable `RUBIES` is defined in the ERB binding during pipeline file generation. This variable looks like:

``` ruby
  # these numbers/names align with public docker image names
  RUBIES = {
    mri: %w[2.5 2.6 2.7 3.0], # docker repository: "ruby"
    jruby: %w[9.2], # docker repository: "jruby"
    rbx: %w[latest], # docker repository: "rubinius/docker"
    windows: %w[2.3 2.4 2.5 2.6], # windows-ruby-dev-tools-release
    truffle: %w[stable nightly] # docker repository: flavorjones/truffleruby
  }
```

and allows you to write a pipeline like this to get coverage on all the supported rubies:

``` yaml
# myproject.yml
jobs:
% RUBIES[:mri].each do |ruby_version|
  - name: "ruby-<%= ruby_version %>"
    plan:
      ...
      - task: rake-test
        config:
          platform: linux
          image_resource:
            type: docker-image
            source: {repository: "ruby", tag: "<%= ruby_version %>"}
    ...
% end
```

Note that the `windows` rubies are not Docker images, since Concourse's Houdini backend doesn't use Docker. Instead, these are implicitly referring to the supported ruby versions installed by the BOSH release at https://github.com/flavorjones/windows-ruby-dev-tools-release


### YTT Templating

As of v0.39.0, YTT templates are also supported. It's off by default, but you can enable it by passing `ytt: true` to the `Concourse.new` or `Concourse#add_pipeline` calls in your Rakefile (see below for more context).

Your Concourse pipeline configuration file, `<myproject>.yml` (or whatever you've configured with the `:pipeline_erb_filename` parameter), will be treated like a YTT template.

You can optionally specify a directory containing your YTT configuration files (`.star` files and `.yml` templates) so that you can inject project-specific logic.

(If you're unfamiliar with YTT, you can [read about it here](https://get-ytt.io/).)


#### Rubies

These YTT variables are defined by the gem, and can be used in your template:

```starlark
#! ruby.star
cruby_versions = {
    "out_of_support": ["2.4", "2.3", "2.2", "2.1", "2.0.0"],
    "supported": ["2.5", "2.6", "2.7", "3.0"],
    "beta": []
}
jruby_versions = {
    "out_of_support": ["9.1"],
    "supported": ["9.2"],
    "beta": []
}
truffleruby_versions = {
    "supported": ["stable"],
    "out_of_support": [],
    "beta": ["nightly"]
}
```

Here's a simple example:

``` yaml
# myproject.yml

#@ load("ruby.star", "cruby_versions")

---
jobs:
#@ for ruby_version in cruby_versions["supported"]:
  - name: #@ "ruby-{}".format(ruby_version)
    plan:
      ...
      - task: rake-test
        config:
          platform: linux
          image_resource:
            type: docker-image
            source:
              repository: "ruby"
              tag: #@ "mri-{}".format(ruby_version)
    ...
#@ end
```


### Mixing ERB and YTT Templating

Why would you do this? Well, if you really want to, I'll tell you a secret: we treat the YTT templates as though they're *also* ERB templates. So go crazy.


### Secrets

You can use a separate file to keep your pipeline variables secret. By default, `concourse/private.yml` will be used. This filename can also be configured (see below)

If the secrets file exists, it will be passed to the `fly` commandline with the `-l` option to fill in template values.

For example, I might have a concourse config that looks like this:

``` yaml
  - name: nokogiri-pr
    type: pull-request
    source:
      repo: sparklemotion/nokogiri
      access_token: {{github-repo-status-access-token}}
      ignore_paths:
        - concourse/**
```

I can put my access token in `private.yml` like this:

``` yaml
github-repo-status-access-token: "your-token-here"
```

and the final generate template will substitute your credentials into the appropriate place.


### Multiple pipelines

Setting up multiple pipelines might be useful for you, for example one pipeline for running tests against `master` and another for pull requests; or a pipeline for tests and a pipeline for deployments.

If you'd like to set up multiple pipelines, invoke Concourse in your Rakefile like this:

``` ruby
require 'concourse'

Concourse.new("myproject") do |c|
  c.add_pipeline "master", "myproject.yml"
  c.add_pipeline "pull-requests", "pr.yml"
  c.add_pipeline "deploy", "big-red-button.yml"
end
```

Note that when you use the block form:

* it's not necessary to explicitly call `#create_tasks!`
* only the pipelines declared via `#add_pipeline` will be managed

Note also that `Concourse#add_pipeline` takes additional options:

* `ytt`: same as the global `ytt` configuration below


## Configuration

### `directory`: Concourse subdirectory name

You can choose a directory name other than the default `concourse`:

``` ruby
Concourse.new("myproject", directory: "ci").create_tasks!
```


### `fly_target`: Concourse `fly` target name

If the initializer is given no additional parameters, your fly target is assumed to be named "default". But this is likely an inappropriate default, and so you can specify your target name:

``` ruby
Concourse.new("myproject", fly_target: "myci").create_tasks! # `fly -t myci <command>`
```


### `ytt`: Enable YTT templating

By default, this option is `false` and pipelines are treated as ERB templates.

Setting this to `true` will run the file through first ERB (which will be a no-op if you're not using ERB) and then through YTT.

Setting this to a `String` argument containing a directory path will cause YTT to be invoked with a `-f` option loading that directory.


### `format`: Emit the final pipelines in `fly format-pipeline` canonical format

If you'd prefer to have your final pipeline files in `fly`'s "canonical format" (via `format-pipeline`), then set this to true!

``` ruby
Concourse.new("myproject", format: true).create_tasks!
```

This might be useful if you're heavily refactoring your template, and want to make sure there aren't unexpected changes to the pipeline.


### `pipeline_erb_filename`: Pipeline filename

By default the pipeline file will be named `<myproject>.yml`, but can be set to something else:

``` ruby
Concourse.new("myproject", pipeline_erb_filename: "pipeline.yml").create_tasks!
```

Note that the generated, final pipeline file is always named `<pipeline_erb_filename>.generated`.



### `secrets_filename`: Secrets filename

You can use a separate file to keep your pipeline variables secret. By default, `concourse/private.yml` will be used. You can specify a different filename:

``` ruby
Concourse.new("myproject", secrets_filename: "secrets.yml").create_tasks!
```


### `fly_args_<fly_command>`: fly command arguments

Rarely, you may need to inject additional commandline arguments into `fly` to get the behavior you want. For example, I wanted to inject `--enable-across-step` into my `validate-pipeline` commands when I started exploring matrix builds via [the `across` step](https://github.com/vito/rfcs/blob/spatial-resources/029-across-step/proposal.md).

You can pass in additional keys named `fly_args_<command-name-with-underscores>` like this:

``` ruby
Concourse.new("myproject", fly_args_validate_pipeline: "--enable-across-step --another-flag")
```

With the above initializer call, whenever the `concourse` gem invokes `validate-pipeline`, it will inject `--enable-across-step --another-flag`.


## Rake Tasks


### Manage your Concourse pipeline

Tasks to manage a local pipeline file, generated from an ERB template:

```
rake concourse:clean     # remove generated pipeline files
rake concourse:generate  # generate and validate all pipeline files
```

A task to update your pipeline configuration:

```
rake concourse:set       # upload all pipeline files
```

Tasks to publicly expose or hide your pipeline:

```
rake concourse:expose    # expose all pipelines
rake concourse:hide      # hide all pipelines
```

Tasks to pause and unpause your pipeline:

```
rake concourse:pause     # pause all pipelines
rake concourse:unpause   # unpause all pipelines
```

And, should you ever need to [nuke the site from orbit][ripley], a task to destroy your pipeline:

```
rake concourse:destroy   # destroy all pipelines
```

  [ripley]: https://www.youtube.com/watch?v=aCbfMkh940Q


### List pipeline tasks

You can see a list of all the tasks defined in your pipelines:

```
rake concourse:tasks  # list all available tasks from all pipelines
```

You'll see output suitable for the `concourse:task` rake task, formatted as `job-name/task-name`, looking something like:

```
RUNNING: (in /home/flavorjones/code/oss/nokogiri) fly validate-pipeline -c concourse/nokogiri.yml.generated
looks good
NOTE: Available Concourse tasks for nokogiri are:
 * jruby-9.1/rake-test
 * jruby-9.2-pr/rake-test
 * jruby-9.2/rake-test
 * ruby-2.5-system/rake-test
 * ruby-2.5-vendored/rake-test
 * ruby-2.6-system-pr/rake-test
 * ruby-2.6-system/rake-test
 * ruby-2.6-vendored-pr/rake-test
 * ruby-2.6-vendored/rake-test
```


### Run a Concourse task with `fly execute`

You can `fly execute` a task defined in your pipelines:

```
rake concourse:task[job_task,fly_execute_args]  # fly execute the specified task
```

where:

* _required_: `job_task` is formatted as `job-name/task-name`, for example, `ruby-2.7/rake-test`. (See `concourse:tasks` for a list of all available task names.)
* _optional_: `fly_execute_args` will default to map the project directory to a resource with the project name, e.g. `--input=myproject=.`, so your pipeline must name the input resource appropriately in order to use the default.


### Abort running builds

```
rake concourse:abort-builds  # abort all running builds for this pipeline
```


### Prune stalled concourse workers

Especially useful if you're deploying via BOSH, which often results in stalled workers;

```
rake concourse:prune-stalled-workers  # prune any stalled workers
```


## Running a local ephemeral Concourse

You can run a local Concourse deployment on your development system, if you have `docker` and `docker-compose` installed.


### Spinning up a local deployment

To spin up the local cluster:

``` sh
rake concourse:local:up
```

You can view that Concourse deployment at `http://127.0.0.1:8080` using the credentials `test`/`test`.


### Doing things with the local deployment

To target that local cluster, simply prepend the `concourse:local` task on the command line. For example, to `fly execute` a task on the local cluster:

``` sh
rake concourse:local concourse:task[ruby-2.7/rake-task]
```

Or to push your pipelines to that local cluster:

``` sh
rake concourse:local concourse:set
```

### Shutting down the local deployment

Finally, to spin down the cluster (when you're done with it):

``` sh
rake concourse:local:down
```


## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/flavorjones/concourse-gem. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [Contributor Covenant](http://contributor-covenant.org) code of conduct. See [`CODE_OF_CONDUCT.md`](CODE_OF_CONDUCT.md).


## License

The gem is available as open source under the terms of the [MIT License](http://opensource.org/licenses/MIT). See [`LICENSE.txt`](LICENSE.txt).


CHANGELOG.md
# concourse-gem changelog

## v0.41.0 / 2020-12-28

* add ruby 3.0 supported
* remove ruby 3.0-rc beta


## v0.40.0 / 2020-12-19

* remove ruby 2.4, which has reached EOL


## v0.39.0 / 2020-11-16

* add support for [YTT](https://get-ytt.io/) pipeline templates


## v0.38.0 / 2020-10-01

* add ruby 3.0-rc


## v0.37.0 / 2020-10-01

* add TruffleRuby ("truffle") as a known Ruby


## v0.36.0 / 2020-09-07

* remove generated pipelines from rake's "CLOBBER" global (generated pipelines have been recommended for checkin to source control since v0.26.0)


## v0.35.0 / 2020-09-07

* remove "clean" rake task (generated pipelines have been recommended for checkin to source control since v0.26.0)


## v0.34.0 / 2020-08-31

* allow injection of fly command arguments via `fly_args_<command_name>`


## v0.33.0 / 2020-03-31

* add `format` option for formatting final pipeline files with `fly format-pipeline`


## v0.32.0 / 2020-03-29

* avoid deprecation warning for `Bundler.with_clean_env`


## v0.31.0 / 2020-01-11

* remove jruby 9.1 (which corresponds to ruby 2.3)


## v0.30.0 / 2020-01-11

* remove ruby 2.3, which has reached EOL
* change ruby 2.7-rc to 2.7 (final)


## v0.29.0 / 2019-10-28

* add `2.7-rc` to the set of MRI Rubies


## v0.28.0 / 2019-08-06

* update the `concourse/concourse` docker image when a new docker-compose file is downloaded, to ensure the two stay in sync.


## v0.27.0 / 2019-04-28

### Features

* Provide description for the `concourse:local` rake task.
* `concourse:task` resolves private variables.


### Bug fixes

* Update error message for `concourse:task` task to match actual args.


## v0.26.0 / 2019-02-02

### Breaking changes

* Generated pipeline files are no longer `.gitignore`d.


## v0.25.0 / 2019-02-02

### Features

* Support running a local ephemeral Concourse cluster with `docker-compose`.


## v0.24.0 / 2019-01-24

### Features

* `$LOAD_PATH` includes the concourse directory, so that `require` will find relative filepaths.
* `erbify_file` method may be called to recursively include yaml erb templates.


### Breaking changes

* `#erbify` has been removed in favor of `#erbify_file`


## v0.23.0 / 2019-01-19

### Features

* Support for multiple pipelines.


## v0.22.0 / 2019-01-18

### Breaking changes

* The name of the generated, final pipeline file is now `<pipeline_erb_filename>.generated`.


### Features

* Introduce rake task `concourse:prune-stalled-workers`
* `Concourse.new` now takes an optional `:fly_target` named param to avoid passing this to each task. [#2] (Thanks, @ebmeierj!)
* `Concourse.new` now takes an optional `:pipeline_erb_filename` named param to set the name of the pipeline file used. [#1] (Thanks, @ebmeierj!)
* `Concourse.new` now takes an optional `:secrets_filename` named param to set the name of the private variable file used. [#1] (Thanks, @ebmeierj!)


## v0.21.0 / 2018-12-26

* Introduce ruby 2.6 final (and remove 2.6-rc)


## v0.20.0 / 2018-11-09

* Introduce ruby 2.6-rc
* Introduce helper methods `production_rubies` and `rc_rubies`


## v0.19.0 / 2018-09-15

Several version of Ruby have reached EOL and have been removed from `RUBIES`:

- ruby 2.1: https://www.ruby-lang.org/en/news/2017/04/01/support-of-ruby-2-1-has-ended/
- ruby 2.2: https://www.ruby-lang.org/en/news/2018/06/20/support-of-ruby-2-2-has-ended/
- jruby 1.7: https://github.com/jruby/jruby/issues/4112


## v0.18.0 / 2018-02-18

* Better default for `input` args to `concourse:task`: use the local directory for all the inputs. This matches the pattern where two different resources are used for ci scripts and the source under test.


## v0.17.0 / 2017-09-24

* Add support for Ruby 2.5 (in general and on Windows)
* Remove `fly` `-x` argument to work with Concourse 3.7.0+


## v0.16.0 / 2017-09-24

Bugfix: require "tempfile", we've been relying on the class being implicitly loaded


## v0.15.0 / 2017-09-24

* Add awareness of Windows Rubies (implicitly linked with https://github.com/flavorjones/windows-ruby-dev-tools-release)


## v0.14.0 / 2017-04-23

* Removed badge functionality. IMHO: too complicated, not useful
* Restoring support for Ruby 1.9 (for JRuby 1.7), ugh.


## v0.13.0 / 2017-03-23

### Features

* allow setting of concourse directory name


## v0.12.0 / 2017-02-14

### Features

* rake task `concourse:init`

### Bug fixes

* properly require 'yaml'


## v0.11.0 / 2017-02-14

Stop using `myproject.yml.erb` and just interpret `myproject.yml` as an ERB template. @jtarchie was right. He's always right.


## v0.10.0 / 2017-02-08

Add `destroy` and `abort-builds` tasks.


## v0.9.0 / 2017-02-08

Bugfix: run `fly execute` with a clean Bundler environment, to avoid accidentally injecting our environment variables into the running task.


## v0.8.0 / 2017-02-08

Expose `#erbify` and `#markdown_badge` methods.


## v0.7.0 / 2017-02-08

Added a rake task to generate badges markdown.


## v0.6.{0,1} / 2017-01-26

If it exists, use `concourse/private.yml` to fill in template values.


## v0.5.0 / 2017-01-23

Renamed the `concourse:tasks` task name arg, and improved the README.


## v0.4.0 / 2017-01-22

Now uses the project name as the `fly execute` input resource name.


## v0.3.0 / 2017-01-22

Avoid depending on `rake/clean`


## v0.2.0 / 2017-01-22

Always regenerate pipeline file.


## v0.1.0 / 2017-01-22

First release.


spec/spec_helper.rb
require "bundler/setup"
require "concourse"

module RspecExampleHelpers
  def in_tmp_dir
    Dir.mktmpdir do |dir|
      Dir.chdir(dir) do
        yield
      end
    end
  end

  def in_assets_dir dirname
    target = File.join("spec", "assets", dirname)
    Dir.chdir(target) do
      yield
    end
  end

  def shush_stdout &block
    old_stdout = $stdout
    $stdout = StringIO.new
    yield
    $stdout = old_stdout
  end
end

RSpec.configure do |config|
  # Enable flags like --only-failures and --next-failure
  config.example_status_persistence_file_path = ".rspec_status"

  config.expect_with :rspec do |c|
    c.syntax = :expect
  end

  config.include RspecExampleHelpers
end


spec/concourse_spec.rb
require "spec_helper"

describe Concourse do
  describe ".new" do
    describe "#project name" do
      it "requires project name" do
        expect { Concourse.new }.to raise_exception(ArgumentError)
      end

      it "is saved" do
        expect(Concourse.new("myproject").project_name).to eq "myproject"
      end
    end
  end

  describe ".new with a block" do
    it "doesn't create a default pipeline" do
      concourse = Concourse.new("myproject") do
        # do nothing
      end

      expect(concourse.pipelines).to be_empty
    end

    it "it yields self" do
      actual_result = nil
      concourse = Concourse.new("myproject") do |object|
        actual_result = object
      end

      expect(concourse).to eq(actual_result)
    end
  end

  describe "#directory" do
    it "defaults to 'concourse'" do
      expect(Concourse.new("myproject").directory).to eq "concourse"
    end

    it "optionally accepts a directory name" do
      concourse = Concourse.new("myproject", directory: "ci")
      expect(concourse.directory).to eq("ci")
    end
  end

  describe "#fly_target" do
    it "defaults to 'default'" do
      expect(Concourse.new("myproject").fly_target).to eq "default"
    end

    it "optionally accepts a fly_target name" do
      expect(Concourse.new("myproject", fly_target: "myci").fly_target).to eq "myci"
    end
  end

  describe "#format" do
    it "default to false'" do
      expect(Concourse.new("myproject").format).to be_falsey
    end

    it "optionally accepts a format boolean" do
      expect(Concourse.new("myproject", format: true).format).to eq(true)
      expect(Concourse.new("myproject", format: false).format).to eq(false)
    end
  end

  describe "#pipelines" do
    context "by default" do
      it "has one pipeline named after the project name" do
        concourse = Concourse.new("myproject")
        expect(concourse.pipelines.length).to eq(1)
        expect(concourse.pipelines.first.name).to eq("myproject")
        expect(concourse.pipelines.first.erb_filename).to eq("concourse/myproject.yml")
        expect(concourse.pipelines.first.filename).to eq("concourse/myproject.yml.generated")
      end
    end
  end

  describe "#secrets_filename" do
    it "is `private.yml` by default" do
      expect(Concourse.new("myproject").secrets_filename).to eq("concourse/private.yml")
    end

    it "can be set" do
      expect(Concourse.new("myproject", secrets_filename: "secrets.yml").secrets_filename).to eq("concourse/secrets.yml")
    end
  end

  describe "#add_pipeline" do
    it "requires name and filename arguments" do
      expect { Concourse.new("myproject").add_pipeline("a") }.to raise_exception(ArgumentError)
    end

    it "creates a Pipeline and adds it to the #pipelines array" do
      concourse = Concourse.new("myproject") do |c|
        c.add_pipeline "foo", "fizzle.yml"
        c.add_pipeline "bar", "fozzle.yml"
      end

      expect(concourse.pipelines.length).to eq(2)
      concourse.pipelines.first.tap do |pipeline|
        expect(pipeline.name).to eq("foo")
        expect(pipeline.directory).to eq(concourse.directory)
        expect(pipeline.erb_filename).to eq("concourse/fizzle.yml")
      end
      concourse.pipelines.last.tap do |pipeline|
        expect(pipeline.name).to eq("bar")
        expect(pipeline.directory).to eq(concourse.directory)
        expect(pipeline.erb_filename).to eq("concourse/fozzle.yml")
      end
    end
  end
end


lib/concourse.rb
require "concourse/version"
require "concourse/util"
require "concourse/pipeline"
require "yaml"
require "tempfile"
require "open-uri"

class Concourse
  include Rake::DSL
  include Concourse::Util

  # these numbers/names align with public docker image names
  RUBIES = {
    mri: %w[2.5 2.6 2.7 3.0], # docker repository: "ruby"
    jruby: %w[9.2], # docker repository: "jruby"
    rbx: %w[latest], # docker repository: "rubinius/docker"
    windows: %w[2.3 2.4 2.5 2.6], # windows-ruby-dev-tools-release
    truffle: %w[stable nightly] # docker repository: flavorjones/truffleruby
  }

  DEFAULT_DIRECTORY = "concourse"
  DEFAULT_FLY_TARGET = "default"
  DEFAULT_SECRETS = "private.yml"

  attr_reader :project_name
  attr_reader :directory
  attr_reader :pipelines
  attr_reader :fly_target
  attr_reader :fly_args
  attr_reader :secrets_filename
  attr_reader :format
  attr_reader :ytt

  CONCOURSE_DOCKER_COMPOSE = "docker-compose.yml"

  def self.url_for(fly_target)
    matching_line = `fly targets`.split("\n").grep(/^#{fly_target}/).first
    raise "invalid fly target #{fly_target}" unless matching_line
    matching_line.split(/ +/)[1]
  end

  def self.default_execute_args(task)
    args = []
    task["config"]["inputs"].each do |input|
      args << "--input=#{input["name"]}=."
    end
    args.join(" ")
  end

  def self.production_rubies
    RUBIES[:mri].reject { |r| r =~ /rc/ }
  end

  def self.rc_rubies
    RUBIES[:mri].select { |r| r =~ /rc/ }
  end

  def initialize(project_name, options = {}, &block)
    @project_name = project_name

    @directory = options[:directory] || DEFAULT_DIRECTORY
    @fly_target = options[:fly_target] || DEFAULT_FLY_TARGET
    @format = options.has_key?(:format) ? options[:format] : false
    @ytt = options.has_key?(:ytt) ? options[:ytt] : false
    @fly_args = options.keys.grep(/^fly_args_/).inject({}) do |hash, key|
      fly_command = key.to_s.gsub(/^fly_args_/, "").gsub("_", "-")
      hash[fly_command] = options[key]
      hash
    end

    base_secrets_filename = options[:secrets_filename] || DEFAULT_SECRETS
    @secrets_filename = File.join(@directory, base_secrets_filename)

    @pipelines = []
    if block
      block.call(self)
      create_tasks!
    else
      add_pipeline(@project_name, (options[:pipeline_erb_filename] || "#{project_name}.yml"), {ytt: ytt})
    end
  end

  def add_pipeline(name, erb_filename, options={})
    @pipelines << Concourse::Pipeline.new(name, @directory, erb_filename, options)
  end

  def pipeline_subcommands(command)
    pipelines.collect { |p| "#{command}:#{p.name}" }
  end

  def rake_init
    FileUtils.mkdir_p File.join(directory, "tasks")
    pipelines.each do |pipeline|
      FileUtils.touch pipeline.erb_filename
    end
    ensure_in_gitignore secrets_filename
  end

  def rake_pipeline_generate(pipeline)
    pipeline.generate
    fly "validate-pipeline", "-c #{pipeline.filename}"
    fly "format-pipeline", "-c #{pipeline.filename} -w" if format
  end

  def ensure_docker_compose_file
    return if File.exist?(docker_compose_path)
    note "fetching docker compose file ..."
    File.open(docker_compose_path, "w") do |f|
      f.write URI.open("https://concourse-ci.org/docker-compose.yml").read
      sh "docker pull concourse/concourse"
    end
  end

  def rake_concourse_local
    ensure_docker_compose_file
    @fly_target = "local"
    fly "login", "-u test -p test -c http://127.0.0.1:8080"
  end

  def rake_concourse_local_up
    ensure_docker_compose_file
    docker_compose "up -d"
    docker_compose "ps"
  end

  def rake_concourse_local_down
    ensure_docker_compose_file
    docker_compose "down"
  end

  def create_tasks!
    unless Dir.exist? directory
      mkdir_p directory
    end

    pipelines.each do |pipeline|
      unless File.exist? pipeline.erb_filename
        warn "WARNING: concourse template #{pipeline.erb_filename.inspect} does not exist, run `rake concourse:init`"
      end
    end

    namespace :concourse do
      #
      #  project commands
      #
      desc "bootstrap a concourse config"
      task :init do
        rake_init
      end

      #
      #  pipeline commands
      #
      desc "generate and validate all pipeline files"
      task "generate" => pipeline_subcommands("generate")

      pipelines.each do |pipeline|
        desc "generate and validate the #{pipeline.name} pipeline file"
        task "generate:#{pipeline.name}" do
          rake_pipeline_generate pipeline
        end
      end

      desc "upload all pipeline files"
      task "set" => pipeline_subcommands("set")

      pipelines.each do |pipeline|
        desc "upload the #{pipeline.name} pipeline file"
        task "set:#{pipeline.name}" => "generate:#{pipeline.name}" do
          options = [
            "-p '#{pipeline.name}'",
            "-c '#{pipeline.filename}'",
          ]
          if File.exist? secrets_filename
            note "using #{secrets_filename} to resolve template vars in #{pipeline.filename}"
            options << "-l '#{secrets_filename}'"
          end
          fly "set-pipeline", options.join(" ")
        end
      end

      %w[expose hide pause unpause destroy].each do |command|
        desc "#{command} all pipelines"
        task command => pipeline_subcommands(command)

        pipelines.each do |pipeline|
          desc "#{command} the #{pipeline.name} pipeline"
          task "#{command}:#{pipeline.name}" do
            fly "#{command}-pipeline", "-p #{pipeline.name}"
          end
        end
      end

      #
      #  task commands
      #
      desc "list all available tasks from all pipelines"
      task "tasks" => "generate" do
        tasks = []

        pipelines.each do |pipeline|
          each_task(pipeline) do |job, task|
            tasks << "#{job["name"]}/#{task["task"]}"
          end
        end

        note "Available Concourse tasks for #{project_name} are:"
        tasks.sort.each { |task| puts " * #{task}" }
      end

      desc "fly execute the specified task"
      task "task", [:job_task, :fly_execute_args] => "generate" do |t, args|
        job_task = args[:job_task]
        unless job_task
          raise "ERROR: must specify a task name, like `rake #{t.name}[taskname]`"
        end

        concourse_task = find_task(job_task)
        raise "ERROR: could not find task `#{job_task}`" unless concourse_task

        fly_execute_args = args[:fly_execute_args] || Concourse.default_execute_args(concourse_task)

        if File.exist? secrets_filename
          note "using #{secrets_filename} to resolve template vars"
          fly_execute_args += " -l '#{secrets_filename}'"
        end

        puts concourse_task.to_yaml

        Tempfile.create("concourse-task") do |f|
          f.write concourse_task["config"].to_yaml
          f.close
          Bundler.with_unbundled_env do
            fly "execute", [fly_execute_args, "-c #{f.path}"].compact.join(" ")
          end
        end
      end

      #
      #  builds commands
      #
      desc "abort all running builds for this concourse team"
      task "abort-builds" do |t, args|
        `fly -t #{fly_target} builds`.each_line do |line|
          pipeline_job, build_id, status = *line.split(/\s+/)[1, 3]
          next unless status == "started"

          fly "abort-build", "-j #{pipeline_job} -b #{build_id}"
        end
      end

      #
      #  worker commands
      #
      desc "prune any stalled workers"
      task "prune-stalled-workers" do
        `fly -t #{fly_target} workers | fgrep stalled`.each_line do |line|
          worker_id = line.split.first
          fly "prune-worker", "-w #{worker_id}"
        end
      end

      #
      #  docker commands
      #
      desc "set fly target to the local docker-compose cluster"
      task "local" do
        rake_concourse_local
      end

      namespace "local" do
        desc "start up a docker-compose cluster for local CI"
        task "up" do
          rake_concourse_local_up
        end

        desc "shut down the docker-compose cluster"
        task "down" do
          rake_concourse_local_down
        end
      end
    end
  end
end
